File: 2020/4/main.py
#!/usr/bin/env python3

def get_passport(input):
    passport = ""

    for x in input:
        if x == '\n':
            yield passport
            passport = ""
        passport += x
    
    yield passport

def is_valid(passport):
    required_fields = ["byr:", "iyr:", "eyr:", "hgt:", "hcl:", "ecl:", "pid:"]

    if all(x in passport for x in required_fields):
        passport = passport.replace("\n", " ")
        for x in required_fields:
            field_data_start = passport.find(x) + len(x)
            field_data_end = passport[field_data_start:].find(" ")
            field_data = passport[field_data_start:field_data_start+field_data_end+1].strip()

            if x == "byr:":
                if len(field_data) > 4:
                    return False
                try:
                    data = int(field_data)
                except:
                    return False

                if data < 1920 or data > 2002:
                    return False 
                
            if x == "iyr:":
                if len(field_data) > 4:
                    return False
                data = int(field_data)

                if data < 2010 or data > 2020:
                    return False 
            if x == "eyr:":
                if len(field_data) > 4:
                    return False
                data = int(field_data)

                if data < 2020 or data > 2030:
                    return False 
            if x == "hgt:":
                if not "cm" in field_data and not "in" in field_data:
                    return False
                
                height = int(field_data[:len(field_data) - 2])

                if "cm" in field_data and (height < 150 or height > 193):
                    return False
                if "in" in field_data and (height < 59 or height > 76):
                    return False
            if x == "hcl:":
                valid = "0123456789abcdef"
                if field_data[0] != '#':
                    return False
                
                for z in field_data[1:]:
                    if z not in valid:
                        return False
                if len(field_data) != 7:
                    return False
                pass
            if x == "ecl:":
                valid = ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]
                if not field_data in valid:
                    return False
            if x == "pid:":
                if len(field_data) != 9:
                    return False
                try:
                    data = int(field_data)
                except:
                    return False
        
        return True
    return False


def main():
    input = open("input.txt", 'r').readlines()

    valid = 0
    for x in get_passport(input):
        if (is_valid(x)):
            valid += 1
    
    print("Valid:", valid)



main()

File: 2020/4/src/main.rs
use std::fs;
#[macro_use]
extern crate lazy_static;
extern crate regex;
use regex::Regex;

fn in_range(num: i32, lower: i32, upper: i32) -> bool {
    return lower <= num && num <= upper;
}

fn is_valid(passport: &str) -> bool {
    lazy_static! {
        static ref BYR_RE: Regex = Regex::new(r"byr:\d{4}").unwrap();
        static ref IYR_RE: Regex = Regex::new(r"iyr:\d{4}").unwrap();
        static ref EYR_RE: Regex = Regex::new(r"eyr:\d{4}").unwrap();
        static ref HGT_RE: Regex = Regex::new(r"hgt:\d*(in|cm)").unwrap();
        static ref HCL_RE: Regex = Regex::new(r"hcl:#[0-9a-f]{6}").unwrap();
        static ref EYE_RE: Regex = Regex::new(r"ecl:(amb|blu|brn|gry|grn|hzl|oth)").unwrap();
        static ref PID_RE: Regex = Regex::new(r"pid:\b\d{9}\b").unwrap();
    }

    let validators = [
        match BYR_RE.find(passport) {
            Some(m) => in_range(
                passport[m.start() + 4..m.end()].parse::<i32>().unwrap(),
                1920,
                2002,
            ),
            None => false,
        },
        match IYR_RE.find(passport) {
            Some(m) => in_range(
                passport[m.start() + 4..m.end()].parse::<i32>().unwrap(),
                2010,
                2020,
            ),
            None => false,
        },
        match EYR_RE.find(passport) {
            Some(m) => in_range(
                passport[m.start() + 4..m.end()].parse::<i32>().unwrap(),
                2020,
                2030,
            ),
            None => false,
        },
        match HGT_RE.find(passport) {
            Some(m) => {
                let val = passport[m.start() + 4..m.end() - 2].parse::<i32>().unwrap();
                let unit = &passport[m.end() - 2..m.end()];

                if unit.eq("cm") {
                    in_range(val, 150, 193)
                } else {
                    in_range(val, 59, 76)
                }
            }
            None => false,
        },
        HCL_RE.is_match(passport),
        EYE_RE.is_match(passport),
        PID_RE.is_match(passport),
    ];

    return validators.iter().all(|&x| x == true);
}
fn main() {
    let contents = fs::read_to_string("input.txt").expect("Failed to read input");
    let input_data = contents.split("\n\n");
    // println!("{:#?}", input_data)

    let mut valid = 0;

    for passport in input_data {
        let pass = String::from(passport).replace("\n", " ");

        if is_valid(&pass) {
            valid += 1;
        }
    }

    println!("Valid: {}", valid);
}
